csi/controller_server.go

```diff
@@ -545,6 +547,47 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 		return nil, status.Error(codes.InvalidArgument, "Snapshot name must be provided")
 	}
 
+	volume, err := cs.apiClient.Volume.ById(csiVolumeName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get volume %v", csiVolumeName)
+	}
+	if volume.State == string(longhorn.VolumeStateDetached) {
+		logrus.Debugf("CSI snapshot attempting to attach %v volume to a random node", volume.Name)
+		nodeID, err := cs.autoAttachVolumeToRandomNode(volume)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		logrus.Debugf("CSI snapshot attached %v volume to %v node", volume.Name, nodeID)
+
+		defer func() {
+			volume, err := cs.apiClient.Volume.ById(csiVolumeName)
+			if err != nil {
+				logrus.WithError(err).Errorf("volume %v unavailable", csiVolumeName)
+				return
+			}
+			if volume == nil {
+				logrus.WithError(err).Errorf("volume %v not found", csiVolumeName)
+				return
+			}
+			logrus.Debugf("CSI snapshot attempting to detach %v volume on %s node", csiVolumeName, nodeID)
+			_, err = cs.unpublishVolume(volume, nodeID, func() error {
```

Comment:
Wait... The volume cannot be detached before the backup creation is **completed**. I think we should use a goroutine rather than a defer function here...

Reply:
Addressed.